File: 2392-successful-pairs-of-spells-and-potions/2392-successful-pairs-of-spells-and-potions.py
class Solution(object):
    def successfulPairs(self, spells, potions, success):
        """
        :type spells: List[int]
        :type potions: List[int]
        :type success: int
        :rtype: List[int]
        """
        potions.sort()
        result = []
        for spell in spells:
            l, r = 0, len(potions)-1
            idx = len(potions)
            while l <= r:
                mid = (l + r) // 2
                if spell * potions[mid] >= success:
                    r = mid-1
                    idx = mid
                else:
                    l = mid+1
            
            result.append(len(potions) - idx)
        
        return result
                
        

File: 2392-successful-pairs-of-spells-and-potions/2392-successful-pairs-of-spells-and-potions.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        List<Integer> list = new ArrayList<>();
        Arrays.sort(potions);

        for (int spell : spells) {
            int left = 0, right = potions.length-1;
            int idx = potions.length;
            while (left <= right) {
                int mid = (left + right) / 2;
                
                if ((long) spell * potions[mid] >= success) {
                    right = mid-1;
                    idx = mid;
                } else {
                    left = mid+1;
                }
            }
            list.add(potions.length - idx);
        }

        return list.stream().mapToInt(Integer::intValue).toArray();
    }
}